File: runtime/ts/src/dispatcher-ownership.test.ts
import assert from 'node:assert/strict';
import test from 'node:test';
import type { Endpoint, Message, Path, Receiver } from '@nightseam/duplex';
import { mount } from '@nightseam/duplex';
import { createDispatcher } from './dispatcher.ts';
import { wirePair } from './wire-pair.ts';
import { emitWire } from './wire.ts';

const settled = (): Promise<void> => new Promise((resolve) => setTimeout(resolve, 0));
const anEvent = (): Message => ({ frame: { version: 1, kind: 'event', data: null } });

/** An endpoint that hands a test its one attachment, and nothing else. */
class Holder implements Endpoint {
  private attached: Receiver | undefined;
  send(): void {}
  receive(receiver: Receiver): () => void {
    if (this.attached) throw new Error('receiver_exists');
    this.attached = receiver;
    return () => {
      if (this.attached === receiver) this.attached = undefined;
    };
  }
  close(): void {
    this.attached = undefined;
  }
  deliver(path: Path, message: Message): void {
    void this.attached?.message?.(path, message);
  }
}

test('an endpoint has one owning attachment', async () => {
  const [left, right] = wirePair();
  try {
    const first: Path[] = [];
    const second: Path[] = [];
    const detach = right.receive({
      message(path) {
        first.push(path);
      },
    });
    assert.throws(() => right.receive({}), { code: 'receiver_exists' });
    emitWire(left, ['one']);
    await settled();
    assert.deepEqual(first, [['one']]);
    detach();
    detach();
    right.receive({
      message(path) {
        second.push(path);
      },
    });
    emitWire(left, ['two']);
    await settled();
    assert.deepEqual(second, [['two']]);
    assert.deepEqual(first, [['one']], 'the detached attachment still received');
  } finally {
    left.close();
  }
});

test('a dispatcher refuses a duplicate path and frees it on detach', () => {
  const dispatch = createDispatcher(new Holder());
  const detach = dispatch.register(['a', 'b'], { message() {} });
  assert.throws(() => dispatch.register(['a', 'b'], {}), { code: 'receiver_exists' });
  // Exact and prefix are separate spaces: one of each may hold a path.
  dispatch.registerPrefix(['a', 'b'], {});
  assert.throws(() => dispatch.registerPrefix(['a', 'b'], {}), { code: 'receiver_exists' });
  detach();
  dispatch.register(['a', 'b'], {});
});

test('overlapping routes select the longest, then the exact', () => {
  const endpoint = new Holder();
  const dispatch = createDispatcher(endpoint);
  const reached: string[] = [];
  const name = (label: string): Receiver => ({
    message() {
      reached.push(label);
    },
  });
  dispatch.registerPrefix([], name('root'));
  dispatch.registerPrefix(['a'], name('a'));
  dispatch.registerPrefix(['a', 'b'], name('a/b'));
  dispatch.register(['a', 'b', 'c'], name('exact a/b/c'));
  for (const [path, label] of [
    [['z'], 'root'],
    [['a', 'z'], 'a'],
    [['a', 'b', 'z'], 'a/b'],
    [['a', 'b', 'c'], 'exact a/b/c'],
  ] as [Path, string][]) {
    reached.length = 0;
    endpoint.deliver(path, anEvent());
    assert.deepEqual(reached, [label], `${JSON.stringify(path)}`);
  }
});

test('a nested selection prepends outgoing and strips incoming', async () => {
  const [left, right] = wirePair();
  try {
    const dispatch = createDispatcher(right);
    const inner = dispatch.select(['a']).select(['b']);
    const delivered: Path[] = [];
    inner.receive({
      message(path) {
        delivered.push(path);
      },
    });
    emitWire(left, ['a', 'b', 'read']);
    await settled();
    assert.deepEqual(delivered, [['read']]);
    const back: Path[] = [];
    left.receive({
      message(path) {
        back.push(path);
      },
    });
    emitWire(inner, ['reply']);
    await settled();
    assert.deepEqual(back, [['a', 'b', 'reply']]);
  } finally {
    left.close();
  }
});

test('a mount routes by one segment and borrows its children', async () => {
  const [left, right] = wirePair();
  try {
    const root = mount(new Map([['child', right]]));
    const delivered: Path[] = [];
    root.receive({
      message(path) {
        delivered.push(path);
      },
    });
    emitWire(left, ['read']);
    await settled();
    assert.deepEqual(delivered, [['child', 'read']]);
    assert.throws(() => root.send([], anEvent()), { code: 'no_route' });
    assert.throws(() => root.send(['absent'], anEvent()), { code: 'no_route' });
    root.close();
    const after: Path[] = [];
    right.receive({
      message(path) {
        after.push(path);
      },
    });
    emitWire(left, ['again']);
    await settled();
    assert.deepEqual(after, [['again']], 'closing the mount closed its borrowed child');
  } finally {
    left.close();
  }
});

File: runtime/go/dispatcher_ownership_test.go
package runtime_test

import (
	"errors"
	bitwire "github.com/Bitspark/bitwire/wire/go"
	"slices"
	"testing"

	"github.com/Bitspark/nightseam/duplex/go"
	ws "github.com/Bitspark/nightseam/runtime/go"
)

// An endpoint has one owning attachment: a second is refused without replacing
// the first, detach is idempotent, and a later attachment is permitted.
func TestAnEndpointHasOneOwningAttachment(t *testing.T) {
	left, right, err := ws.NewWirePair(ws.Options{})
	if err != nil {
		t.Fatal(err)
	}
	t.Cleanup(func() { _ = left.Close(duplex.CodeNormal, "") })
	first := make(chan []string, 4)
	detach, err := right.Receive(bitwire.Receiver{Message: func(path []string, _ bitwire.Message) { first <- path }})
	if err != nil {
		t.Fatal(err)
	}
	if _, err := right.Receive(bitwire.Receiver{}); !errors.Is(err, duplex.ErrReceiverExists) {
		t.Fatalf("a second attachment was accepted: %v", err)
	}
	if err := ws.EmitWire(t.Context(), left, []string{"one"}, nil); err != nil {
		t.Fatal(err)
	}
	if got := <-first; !slices.Equal(got, []string{"one"}) {
		t.Fatalf("first attachment saw %v", got)
	}
	detach()
	detach()
	second := make(chan []string, 4)
	if _, err := right.Receive(bitwire.Receiver{Message: func(path []string, _ bitwire.Message) { second <- path }}); err != nil {
		t.Fatalf("a later attachment was refused: %v", err)
	}
	if err := ws.EmitWire(t.Context(), left, []string{"two"}, nil); err != nil {
		t.Fatal(err)
	}
	if got := <-second; !slices.Equal(got, []string{"two"}) {
		t.Fatalf("second attachment saw %v", got)
	}
	select {
	case got := <-first:
		t.Fatalf("the detached attachment still received %v", got)
	default:
	}
}

// A dispatcher refuses a duplicate path and frees it when its registration
// detaches. Exact and prefix are separate spaces: one of each may hold a path.
func TestADispatcherRefusesADuplicatePath(t *testing.T) {
	endpoint := newInvocationEndpoint(ws.DefaultInvocationLimits())
	dispatch, err := ws.NewDispatcher(endpoint)
	if err != nil {
		t.Fatal(err)
	}
	detach, err := dispatch.Register([]string{"a", "b"}, bitwire.Receiver{Message: func([]string, bitwire.Message) {}})
	if err != nil {
		t.Fatal(err)
	}
	if _, err := dispatch.Register([]string{"a", "b"}, bitwire.Receiver{}); !errors.Is(err, duplex.ErrReceiverExists) {
		t.Fatalf("a duplicate exact path was accepted: %v", err)
	}
	if _, err := dispatch.RegisterPrefix([]string{"a", "b"}, bitwire.Receiver{}); err != nil {
		t.Fatalf("a prefix at an exact path was refused: %v", err)
	}
	if _, err := dispatch.RegisterPrefix([]string{"a", "b"}, bitwire.Receiver{}); !errors.Is(err, duplex.ErrReceiverExists) {
		t.Fatalf("a duplicate prefix path was accepted: %v", err)
	}
	detach()
	if _, err := dispatch.Register([]string{"a", "b"}, bitwire.Receiver{}); err != nil {
		t.Fatalf("a detached path was not freed: %v", err)
	}
}

// Overlapping prefixes: the longest match wins, and an exact registration wins
// over every prefix that would also have matched.
func TestOverlappingRoutesSelectTheLongestThenTheExact(t *testing.T) {
	endpoint := newInvocationEndpoint(ws.DefaultInvocationLimits())
	dispatch, err := ws.NewDispatcher(endpoint)
	if err != nil {
		t.Fatal(err)
	}
	reached := make(chan string, 8)
	name := func(label string) bitwire.Receiver {
		return bitwire.Receiver{Message: func([]string, bitwire.Message) { reached <- label }}
	}
	for _, route := range []struct {
		path  []string
		label string
	}{{nil, "root"}, {[]string{"a"}, "a"}, {[]string{"a", "b"}, "a/b"}} {
		if _, err := dispatch.RegisterPrefix(route.path, name(route.label)); err != nil {
			t.Fatal(err)
		}
	}
	if _, err := dispatch.Register([]string{"a", "b", "c"}, name("exact a/b/c")); err != nil {
		t.Fatal(err)
	}
	for _, want := range []struct {
		path  []string
		label string
	}{
		{[]string{"z"}, "root"},
		{[]string{"a", "z"}, "a"},
		{[]string{"a", "b", "z"}, "a/b"},
		{[]string{"a", "b", "c"}, "exact a/b/c"},
	} {
		endpoint.deliver(want.path, bitwire.Message{Frame: bitwire.ProfileFrame{Version: 1, Kind: bitwire.ProfileEvent, Data: []byte("null")}})
		if got := <-reached; got != want.label {
			t.Fatalf("%v reached %q, want %q", want.path, got, want.label)
		}
	}
}

// A nested selection prepends its prefixes outgoing and strips them incoming,
// and every view is a view of the one root attachment.
func TestNestedSelectionPrependsAndStrips(t *testing.T) {
	left, right, err := ws.NewWirePair(ws.Options{})
	if err != nil {
		t.Fatal(err)
	}
	t.Cleanup(func() { _ = left.Close(duplex.CodeNormal, "") })
	dispatch, err := ws.NewDispatcher(right)
	if err != nil {
		t.Fatal(err)
	}
	inner := dispatch.Select([]string{"a"}).Select([]string{"b"})
	delivered := make(chan []string, 4)
	if _, err := inner.Receive(bitwire.Receiver{Message: func(path []string, _ bitwire.Message) { delivered <- path }}); err != nil {
		t.Fatal(err)
	}
	if err := ws.EmitWire(t.Context(), left, []string{"a", "b", "read"}, nil); err != nil {
		t.Fatal(err)
	}
	if got := <-delivered; !slices.Equal(got, []string{"read"}) {
		t.Fatalf("the nested view was delivered %v", got)
	}
	// And outgoing: what the view sends arrives at the root's full path.
	back := make(chan []string, 4)
	if _, err := left.Receive(bitwire.Receiver{Message: func(path []string, _ bitwire.Message) { back <- path }}); err != nil {
		t.Fatal(err)
	}
	if err := ws.EmitWire(t.Context(), inner, []string{"reply"}, nil); err != nil {
		t.Fatal(err)
	}
	if got := <-back; !slices.Equal(got, []string{"a", "b", "reply"}) {
		t.Fatalf("the nested view sent to %v", got)
	}
}

// Mounting chooses a child by one segment and restores it on delivery;
// closing the mount detaches its own attachments and leaves children usable.
func TestMountRoutesByOneSegmentAndBorrowsItsChildren(t *testing.T) {
	left, right, err := ws.NewWirePair(ws.Options{})
	if err != nil {
		t.Fatal(err)
	}
	t.Cleanup(func() { _ = left.Close(duplex.CodeNormal, "") })
	mount := duplex.Mount(map[string]bitwire.Endpoint{"child": right})
	delivered := make(chan []string, 4)
	if _, err := mount.Receive(bitwire.Receiver{Message: func(path []string, _ bitwire.Message) { delivered <- path }}); err != nil {
		t.Fatal(err)
	}
	if err := ws.EmitWire(t.Context(), left, []string{"read"}, nil); err != nil {
		t.Fatal(err)
	}
	if got := <-delivered; !slices.Equal(got, []string{"child", "read"}) {
		t.Fatalf("the mount delivered %v", got)
	}
	// A mount has no destination at the empty path, and an unknown child has
	// no route at all.
	if err := mount.Send(nil, bitwire.Message{Frame: bitwire.ProfileFrame{Version: 1, Kind: bitwire.ProfileEvent, Data: []byte("null")}}); !errors.Is(err, duplex.ErrNoRoute) {
		t.Fatalf("the mount had a destination at []: %v", err)
	}
	if err := mount.Send([]string{"absent"}, bitwire.Message{Frame: bitwire.ProfileFrame{Version: 1, Kind: bitwire.ProfileEvent, Data: []byte("null")}}); !errors.Is(err, duplex.ErrNoRoute) {
		t.Fatalf("an unknown child had a route: %v", err)
	}
	if err := mount.Close(duplex.CodeNormal, ""); err != nil {
		t.Fatal(err)
	}
	after := make(chan []string, 4)
	if _, err := right.Receive(bitwire.Receiver{Message: func(path []string, _ bitwire.Message) { after <- path }}); err != nil {
		t.Fatalf("closing the mount closed its borrowed child: %v", err)
	}
	if err := ws.EmitWire(t.Context(), left, []string{"again"}, nil); err != nil {
		t.Fatal(err)
	}
	if got := <-after; !slices.Equal(got, []string{"again"}) {
		t.Fatalf("the borrowed child delivered %v", got)
	}
}
